src/pobox.erl

```diff
@@ -60,35 +105,70 @@
 %% message ordering.
 %% The initial state can be either passive or notify, depending on whether
 %% the user wants to get notifications of new messages as soon as possible.
--spec start_link(name(), max(), 'stack' | 'queue' | keep_old | {mod, module()}) -> {ok, pid()}.
+-spec start_link(name(), max(), stack | queue | keep_old | {mod, module()}) -> {ok, pid()}.
 start_link(Owner, Size, Type) ->
     start_link(Owner, Size, Type, notify).
 
 %% This one is messy because we have two clauses with 4 values, so we look them
 %% up based on guards.
--spec start_link(name(), max(), 'stack' | 'queue' | keep_old | {mod, module()}, 'notify'|'passive') -> {ok, pid()}
+-spec start_link(name(), max(), stack | queue | keep_old | {mod, module()}, notify | passive) -> {ok, pid()}
         ;       (term(), pid(), max(), stack | queue) -> {ok, pid()}.
-start_link(Owner, Size, Type, StateName) when is_pid(Owner);
-                                              is_atom(Owner),
+start_link(Owner, Size, Type, StateName) when ?PROCESS_NAME_GUARD(Owner),
+                                              ?POBOX_START_STATE_GUARD(StateName),
                                               is_integer(Size), Size > 0 ->
-    gen_statem:start_link(?MODULE, {Owner, Size, Type, StateName}, []);
-start_link(Name, Owner, Size, Type) ->
+    gen_statem:start_link(?MODULE, #pobox_opts{owner = Owner, size = Size, type = Type, state_name = StateName}, []);
+start_link(Name, Owner, Size, Type)
+  when Size > 0,
+    ?PROCESS_NAME_GUARD_WITH_LOCAL_NO_PID(Name),
+    ?PROCESS_NAME_GUARD(Owner),
+    ?POBOX_BUFFER_TYPE_GUARD(Type) ->
     start_link(Name, Owner, Size, Type, notify).
 
--spec start_link(name(), pid(), max(), stack | queue | keep_old | {mod, module()},
+-spec start_link(name(), name(), max(), stack | queue | keep_old | {mod, module()},
```

Comment:
Here I opted to use `name()` for owner instead of using `pid()` because I want to preserve calling semantics if the parent decides he wants this.

Reply:
Switched back to `pid()` for now.